src/nvim/os/time.c

```diff
@@ -46,26 +47,46 @@ void os_delay(uint64_t milliseconds, bool ignoreinput)
     }
     LOOP_PROCESS_EVENTS_UNTIL(&main_loop, NULL, (int)milliseconds, got_int);
   } else {
-    os_microdelay(milliseconds * 1000);
+    os_microdelay(milliseconds * 1000u, ignoreinput);
   }
 }
 
-/// Sleeps for a certain amount of microseconds
+/// Sleeps for a certain amount of microseconds.
 ///
 /// @param microseconds Number of microseconds to sleep
-void os_microdelay(uint64_t microseconds)
+/// @param ignoreinput If true, ignore all input (including SIGINT/CTRL-C).
+///                    If false, waiting is aborted on any input.
+void os_microdelay(uint64_t microseconds, bool ignoreinput)
 {
-  uint64_t elapsed = 0;
-  uint64_t ns = microseconds * 1000;  // convert to nanoseconds
+  uint64_t elapsed = 0u;
   uint64_t base = uv_hrtime();
 
+  // Convert microseconds to nanoseconds. If uint64_t would overflow, set
+  // nanoseconds to UINT64_MAX.
+  const uint64_t nanoseconds = (microseconds < UINT64_MAX/1000u)
+                               ? microseconds * 1000u
+                               : UINT64_MAX;
+
   uv_mutex_lock(&delay_mutex);
 
-  while (elapsed < ns) {
-    if (uv_cond_timedwait(&delay_cond, &delay_mutex, ns - elapsed)
-        == UV_ETIMEDOUT)
+  while (elapsed < nanoseconds) {
+
+    // If the input is ignored, we simply wait the full delay. If not, we
+    // check every 10 milliseconds for input and break the waiting loop if input
+    // is available.
+    const uint64_t nanoseconds_delta = (ignoreinput)
+                                       ? nanoseconds - elapsed
+                                       : MIN(nanoseconds - elapsed, 100000000u);
+    uv_cond_timedwait(&delay_cond, &delay_mutex, nanoseconds_delta);
```

Comment:
I don’t see any benefits of checking the return type here, because we use the return value neither for calculation nor for error checking.

The source code suggests that `uv_cond_timedwait()` can return two values; UV_ETIMEDOUT and 0. Now, my man page says that 0 is the return value in case of success, which doesn’t reflect our results quite well, so let’s first assume that 0 is indeed another value for success, in which case both values were good for us, so the check wouldn’t be necessary. If we, on the other hand, assume that 0 would return in case of failure, then we would have `nanoseconds` of a CPU eating loop (which could be exited in case of `ignoreinput == true`), because currently we don’t use the return value for error handling.

Regarding this, something like
`(void) uv_cond_timedwait(...);`
looks more clear to me than
`const bool expired = uv_cond_timewait(...) == UV_ETIMEDOUT;`
with the second being `expired == false` in case of `uv_cond_timedwait() == 0`. The best solution would be to use the return value for error handling. Can we do that?